### ISSUE TITLE
[solved] Multi-GPUs with Apex FP16

### ISSUE THREAD

**Issue by ngxbac** [catalyst-team/catalyst#209]
Hi, 

I am excited to see `release 19.6`. Especially, I am looking for `FP16` support. 
Thank for your hardworking!
I got a problem where I could not run multi GPUs under `FP16`. 

* Hardware: V100 x 2
* Code: 
```
    runner.train(
        model=model,
        criterion=criterion,
        optimizer=optimizer,
        scheduler=scheduler,
        loaders=loaders,
        logdir=Config.checkpoint,
        num_epochs=Config.epochs,
        verbose=True,
        fp16={"opt_level": "O1"}
    )
```

=> Only one GPU is used 

After looking at the code at: https://github.com/catalyst-team/catalyst/blob/master/catalyst/dl/utils/utils.py#L113 
I tried to change the parameters as: 
```
fp16={"opt_level": "O1", "rank": 0}
```

However, I got this error:
```
ValueError: Error initializing torch.distributed using env:// rendezvous: environment variable RANK expected, but not set
```
I am not familiar with `apex` and `DistributedDataParallel` too much, so if you have any comments, please let me know.

Many thanks

**Comment by ngxbac** [comment]
I hacked the process as following to work with multi GPUs. 
I am not sure the final result is wrong or not.

```
def process_components(
    model: _Model,
    criterion: _Criterion = None,
    optimizer: _Optimizer = None,
    scheduler: _Scheduler = None,
    distributed_params: Dict = None
) -> Tuple[_Model, _Criterion, _Optimizer, _Scheduler, torch.device]:

    distributed_params = distributed_params or {}
    distributed_params = copy.deepcopy(distributed_params)
    device = UtilsFactory.get_device()

    if torch.cuda.is_available():
        cudnn.benchmark = True

    model = model.to(device)

    if is_wrapped_with_ddp(model):
        pass
    elif len(distributed_params) > 0:
        UtilsFactory.assert_fp16_available()
        from apex import amp

        distributed_rank = distributed_params.pop("rank", -1)

        # if distributed_rank > -1:
        #     torch.cuda.set_device(distributed_rank)
        #     torch.distributed.init_process_group(
        #         backend="nccl", init_method="env://")

        model, optimizer = amp.initialize(
            model, optimizer, **distributed_params)

        # if distributed_rank > -1:
            # from apex.parallel import DistributedDataParallel
            # model = DistributedDataParallel(model)
        model = torch.nn.DataParallel(model)
    elif torch.cuda.device_count() > 1:
        model = torch.nn.DataParallel(model)

    model = model.to(device)

    return model, criterion, optimizer, scheduler, device
```

**Comment by Scitator** [comment]
Hi,

So, for now true distributed training with fp16 support is available thought Config API only, an example bash script available on our [19.06 release notes](https://github.com/catalyst-team/catalyst/releases/tag/v19.06)

Nevertheless, your workaround looks good (and should work fine, I think). I suggest you to contribute this solution with `if torch.cuda.device_count() > 1` check if it's not a distributed training case.
